Core/include/Acts/SpacePointFormation/DoubleHitSpacePointBuilder.hpp

```diff
@@ -0,0 +1,129 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2018-2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Units.hpp"
+#include "Acts/Digitization/CartesianSegmentation.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/SpacePointFormation/SpacePointBuilderConfig.h"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/Surfaces/SurfaceArray.hpp"
+
+namespace Acts {
+
+/// @brief Configuration of the class to steer its behaviour
+
+/// @class DoubleHitSpacePointBuilder
+///
+/// After the particle interaction with surfaces are recorded and digitized
+/// the hits strip detectors need further treatment. This class takes
+/// the digitized clusters and combines them on two different detector elements
+/// to a result of the combined detector element. The class is intended to
+/// handle strip detector elements in particular.
+///
+/// @note Used abbreviation: "Strip Detector Element" -> SDE
+///
+
+template <typename spacepoint_t, typename cluster_t>
+class DoubleHitSpacePointBuilder {
+ public:
+  /// Constructor
+  /// @param cfg Specific config that will be used instead of the default values
+  DoubleHitSpacePointBuilder(DoubleHitSpacePointBuilderConfig cfg);
+  ///// Default constructor
+  DoubleHitSpacePointBuilder() = default;
+  /// @brief Searches possible combinations of two clusters on different
+  /// surfaces that may come from the same particles
+  ///
+  /// @param gctx The current geometry context object, e.g. alignment
+  /// @param clustersFront vector of clusters on a surface
+  /// @param clustersBack vector of clusters on another surface
+  /// @param clusterPairs storage of the cluster pairs
+  /// @note The structure of @p clustersFront and @p clustersBack is
+  /// meant to be clusters[Independent clusters on a single surface]
+  void makeClusterPairs(
+      const GeometryContext& gctx,
+      const std::vector<const cluster_t*>& clustersFront,
+      const std::vector<const cluster_t*>& clustersBack,
+      std::vector<std::pair<const cluster_t*, const cluster_t*>>& clusterPairs)
+      const;
+
+  /// @brief Calculates the space points out of a given collection of
+  /// clusters on several strip detectors and stores the data
+  ///
+  /// @param gctx The current geometry context object, e.g. alignment
+  /// @param clusterPairs pairs of clusters that are space point
+  /// candidates
+  /// @param spacePoints storage of the results
+  void calculateSpacePoints(
+      const GeometryContext& gctx,
+      const std::vector<std::pair<const cluster_t*, const cluster_t*>>&
+          clusterPairs,
+      std::vector<spacepoint_t>& spacePoints) const;
```

Comment:
what about introducing also a `calculateSpacePoints` method with the same inputs as in the SingleHitSpacePointBuilder class (thus moving towards a common interface), which internally calls makeClusterPairs and this method? Is this feasible?

```
void calculateSpacePoints(const GeometryContext& gctx,
                            const std::vector<cluster_t>& clusters,
                            std::vector<spacepoint_t>& spacePointStorage) const;
```